File: marketplacebackend/src/main/java/com/ez/marketplacebackend/service/ProductService.java
package com.ez.marketplacebackend.service;

import com.ez.marketplacebackend.entity.Product;
import com.ez.marketplacebackend.repository.ProductRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.persistence.EntityNotFoundException;
import java.util.List;

import static com.ez.marketplacebackend.constant.Constant.NO_PRODUCT_FOUND_BY_ID;

@Service
public class ProductService {

    private Logger LOGGER = LoggerFactory.getLogger(getClass());

    @Autowired
    private ProductRepository productRepository;

    // search products by pageNumber and based on the search criteria.
    // parameters:
    //  - pageNumber: page number
    //  - pageSize: page size
    //  - searchTerm: product name
    public List<Product> searchProducts(int pageNumber, int pageSize, String searchTerm) {

        LOGGER.info("search products");

        return productRepository.searchProducts(pageNumber, pageSize, searchTerm);
    }

    // get total products based on the search criteria
    public long getTotalElements(String searchTerm) {

        LOGGER.info("get total products based on the search criteria");

        return productRepository.getTotalElements(searchTerm);
    }

    // find product by product id
    public Product findById(Long id) throws EntityNotFoundException {

        LOGGER.info("find product by id");

        // find product by product id
        return productRepository.findById(id).orElseThrow(
                () -> new EntityNotFoundException(NO_PRODUCT_FOUND_BY_ID + id));
    }

}

File: marketplacefrontend/src/app/service/product.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { Product } from '../payload/product';

@Injectable({
  providedIn: 'root'
})

export class ProductService {

  // environment.apiUrl: "http://localhost:8080"
  host = environment.apiUrl;

  // inject httpClient
  constructor(
    private httpClient: HttpClient
  ) { }

  // get products by page and based on the search criteria
  searchProducts(pageNumber: number, pageSize: number, searchTerm: string): Observable<Product[]> {

    // ex: url = http://localhost:8080/product-search?pageNumber=0&pageSize=8&searchTerm=
    console.log(`${this.host}/product-search?pageNumber=${pageNumber}&pageSize=${pageSize}&searchTerm=${searchTerm}`)

    return this.httpClient.get<Product[]>(
      `${this.host}/product-search?pageNumber=${pageNumber}&pageSize=${pageSize}&searchTerm=${searchTerm}`
    )

  } // end of searchProducts()

  // calculate total products(total elements)
  getTotalProducts(searchTerm: string): Observable<number> {

    return this.httpClient.get<number>(
      `${this.host}/product-total-elements?searchTerm=${searchTerm}`
    );

  } // end of getTotalProducts()

  // find product by product id
  findById(id: number): Observable<Product> {

    return this.httpClient.get<Product>(`${this.host}/product-list/${id}`);

  } // end of findById()

} // end of class ProductService
